docs/reference/cpp/index.md

```diff
@@ -0,0 +1,254 @@
+
+For most use cases, you can think of C++ as a superset of C. While this is not technically true, more often than not
+you are able to write standard C code for a C++ program without issues. However, doing so ignores a lot of the benefits
+and reasons to use C++.
+
+### Classes and Structs
+
+In C structs can only contain member variables, but in C++ structs are basically classes but with a default member
+visibility of public instead of private.
+
+!!! example
+    The following code blocks are equivalent.
+
+    ```C++
+    struct foo {
+    private:
+        int x;
+        void helper(void);
+    public:
+        foo(int y);
+    }
+    ```
+
+    ```C++
+    class foo {
+    private:
+        int x;
+        void helper(void);
+    public:
+        foo(int y);
+    }
+    ```
+
+### Namespaces
+
+One problem that is prevalent in C concerns the scoping of names. For example, let there be two files `A.h` and `B.h`
+and a program `ighxy.c`, and let them both contain a `float x` and `int bar(void)`.
+
+Our program cannot compile because the linker cannot distinguish which `bar()` function we want to use! One way to fix
+this in a C program would be to rename them `a_bar()` and `b_bar()`. Although this fix seems trivial for this example,
+applying it to a file that has potentially 100 functions can be a lot more difficult, especially if two files just
+happen to share the same prefix for their functions!
+
+C++ introduces namespaces to tackle this problem. With namespaces, we can deal with naming conflicts much more easily.
+Though be aware that namespaces are not necessary everywhere. See the following code snippet to see how they work.
+
+=== "C"
+
+    ```C title="A.h"
+    float x;
+    int bar(void);
+    ```
+
+    ```C title="B.h"
+    float x;
+    int bar(void);
+    ```
+
+    ```C title="ighxy.c"
+    #include "A.h"
+    #include "B.h"
+
+    int main(void) {
+        int a = bar();
+        ...
+    }
+    /* Error, does not compile*/
+    ```
+
+=== "C++"
+
+    ```C++ title="A.h"
+    namespace a {
+    float x;
+    int bar(void);
+    }
+    ```
+
+    ```C++ title="B.h"
+    namespace b {
+    float x;
+    int bar(void);
+    }
+    ```
+
+    ```C++ title="ighxy.cpp"
+    #include "A.h"
+    #include "B.h"
+
+    int main(void) {
+        int a = a::bar();
+        int b = b::bar();
+        float xa = a::x;
+        float xb = b::x;
+        /* No problem! */
+        ...
+    }
+    ```
+
+### Constant Expressions
+
+In C, if we want to declare a constant or a function/expression that we want to be evaluated at compile time, we need
+to use `#define` statements. One of the problems with `#define` statements is that they perform a simple copy paste
+wherever they're used. For example:
+
+=== "Before Precompile"
+
+    ```C
+    #define PI 3.14F
+    #define AREA_OF_CIRCLE(radius) ((PI) * (radius) * (radius))
+
+    int main(void) {
+        float area = AREA_OF_CIRCLE(2.5F);
+        ...
+    }
+    ```
+
```

Comment:
It might be worth it to mention that AREA_OF_CIRCLE is a macro with arguments. You can link them to [this](https://www.cs.yale.edu/homes/aspnes/pinewiki/C(2f)Macros.html) resource that explains C macros nicely.